### DISCUSSION TITLE
Reason for not all types being supported for 2D and 3D matrices?

### DISCUSSION THREAD

**Discussion by itay7564** [beehive-lab/TornadoVM#267] [Ideas/Proposals]
Hi there, hope you are having a nice day!
I was wondering if there is a particular reason for implemented 2d matrix types to be only of float, float4, double, and int
And implemented 3d matrix types are only float and float4.

My guess is it was just not seen as needed for users to have all types implemented, so only the types above were.
The issue is in my project I want to pass a 3D array of bytes, and return a 3D array of Doubles, and I was surprised to see not all types are supported in 2D&3D matrices.

If you tell me I'm correct in my assumption I might as well implement this (rather simple) code and make a PR.

**Comment by jjfumero** [comment]
Hi @itay7564 . 
  Yes, your assumptions are right. No particular reason to not implement those types apart from covering the use cases we have. As a side note, In general, we try to avoid `double` or `fp64` on GPUs, since performance can drop quite a bit, but if someone needs this, of course, we will provide those types. 

As a workaround, you can use array of primitive types (those are fully supported) in 1D, and change accesses to have the "3D" representation within your kernels. 

One thing regarding the TornadoVM types: we are currently changing the APIs to work with the Panama API, so expect in the near future, to have more types supported along with our integration with Panama. 

Regarding contributions, if possible, please hold on while we prepare our open-source contributor agreement. This is not up to us (TornadoVM team), and we are pending from The University of Manchester to provide this (which we expect it will be ready soon).

**Comment by itay7564** [comment]
Thanks for the reply.
